File: golang-backend/internal/handlers/userhandler/dto.go
package userhandler

import "github.com/aportela/doneo/internal/handlers"

// TODO: this struct is duplicated on auth handler
type permissionsFlags struct {
	IsSuperUser bool `json:"isSuperUser"`
}

type addRequest struct {
	Name        string           `json:"name"`
	Email       string           `json:"email"`
	Password    string           `json:"password"`
	Permissions permissionsFlags `json:"permissions"`
}

type updateRequest struct {
	Id          string           `json:"id"`
	Name        string           `json:"name"`
	Email       string           `json:"email"`
	Password    *string          `json:"password,omitempty"`
	Permissions permissionsFlags `json:"permissions"`
}

type patchRequest struct {
	DeletedAt *int64 `json:"deletedAt"`
}

type filterUserPermissions struct {
	IsSuperUser *bool `json:"isSuperUser"`
}

type FilterRequest struct {
	Name        *string                   `json:"name"`
	Email       *string                   `json:"email"`
	Permissions *filterUserPermissions    `json:"permissions"`
	CreatedAt   *handlers.TimestampFilter `json:"createdAt"`
	UpdatedAt   *handlers.TimestampFilter `json:"updatedAt"`
	DeletedAt   *handlers.TimestampFilter `json:"deletedAt"`
}

type searchRequest struct {
	Pager  handlers.PagerRequest `json:"pager"`
	Order  handlers.OrderRequest `json:"order"`
	Filter *FilterRequest        `json:"filter"`
}

type UserBaseResponse struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	AvatarURL string `json:"avatarUrl"`
}

type userResponse struct {
	ID          string           `json:"id"`
	Name        string           `json:"name"`
	Email       string           `json:"email"`
	CreatedAt   int64            `json:"createdAt"`
	UpdatedAt   *int64           `json:"updatedAt"`
	DeletedAt   *int64           `json:"deletedAt"`
	Permissions permissionsFlags `json:"permissions"`
	AvatarURL   string           `json:"avatarUrl"`
}

type searchResponse struct {
	Users []userResponse         `json:"users"`
	Pager handlers.PagerResponse `json:"pager"`
}

File: vite-vue-ts-frontend/src/modules/users/types/dto.ts
import type {
  PagerRequest,
  PagerResponse,
} from "../../../shared/types/dto/pager";
import type { Order } from "../../../shared/types/dto/order";
import type { TimestampRange } from "../../../shared/composables/timestamps";

/*
export type UserRequest = {
  id: string;
  name: string;
  email: string;
  isSuperUser: boolean;
  createdAt: number;
  updatedAt: number | null;
  deletedAt: number | null;
};
*/

interface UserPermissions {
  isSuperUser: boolean;
}

export type AddRequest = {
  name: string;
  email: string;
  permissions: UserPermissions;
};

export type UpdateRequest = {
  id: string;
  name: string;
  email: string;
  password?: string;
  permissions: UserPermissions;
};

interface SearchRequestUserPermissions {
  isSuperUser?: boolean;
}

type SearchRequestFilter = {
  type?: number;
  name?: string;
  email?: string;
  permissions?: SearchRequestUserPermissions;
  createdAt?: TimestampRange;
  updatedAt?: TimestampRange;
  deletedAt?: TimestampRange;
};

export type SearchRequest = {
  pager: PagerRequest;
  order: Order;
  filter?: SearchRequestFilter;
};

export type UserBaseResponse = {
  id: string;
  name: string;
};

export type UserResponse = {
  id: string;
  name: string;
  email: string;
  permissions: UserPermissions;
  createdAt: number;
  updatedAt: number | null;
  deletedAt: number | null;
};

export type SearchBaseResponse = {
  users: UserResponse[];
};

export type SearchResponse = {
  users: UserResponse[];
  pager: PagerResponse;
};
